### DISCUSSION TITLE
Proposal to Compile Performance-Critical Functions for Improved Efficiency

### DISCUSSION THREAD

**Discussion by GabrielJMS** [networkx/networkx#7612] [Ideas]
Hello, NetworkX developers and contributors,

I'm reaching out to discuss the possibility of compiling certain performance-critical functions within the NetworkX codebase to improve efficiency, especially for large-scale graph operations. Given the computational intensity of some graph algorithms, I believe that compiling these functions using Cython or leveraging existing C++ graph libraries with Python bindings could result in significant performance improvements.

### **Potential Areas for Optimization**:
- **Graph Traversal Algorithms**: e.g., BFS, DFS, shortest-path algorithms (like Dijkstra’s, A*, etc.).
- **Clustering and Connectivity Algorithms**.
- **Network Metrics Calculations**: e.g., centrality measures, community detection, etc.

### **Proposed Approaches**:
1. **Cython Compilation**: Identify and compile key functions that could benefit from Cython's speedup capabilities.
2. **C++ Integration**: Integrate C++ implementations of critical algorithms using Pybind11 or similar binding tools.

### **Benefits**:
- Improved performance for large graphs and complex operations.
- Potential for scaling NetworkX to handle even more extensive networks efficiently.
- Retaining the ease of use of NetworkX while enhancing its capability for performance-critical tasks.

### **Questions for the Community**:
- **Interest Level**: Is there interest in pursuing this approach? Would the community benefit from these optimizations?
- **Focus Areas**: Are there specific algorithms or functions that the community has identified as performance bottlenecks that could be targeted first?
- **Collaborators**: Would anyone be interested in collaborating on this initiative?

I'm eager to hear your thoughts and feedback on this idea. If there is interest, I'm happy to help coordinate or contribute to this effort.

Thank you!

**Comment by dschult** [comment]
Historically speaking we have often considered (and sometimes attempted) compiled implementations. But the time needed to convert to the compiled code data structures and/or the reduction in flexibility of node/edge objects were too much. In the last few years we have implemented a backend system that allows compiled functions to be implemented as networkx backend packages. Once installed (and turned on) the backend code is used whenever the specific networkx function is called.  Some example backends already implemented include `graphblas-algorithms`, `nx-cugraph` and `nx-parallel`.  The NetworkX docs for many functions include info on backends that they use e.g. [betweenness_centrality](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.betweenness_centrality.html#networkx.algorithms.centrality.betweenness_centrality) or [has_path](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.generic.has_path.html#networkx.algorithms.shortest_paths.generic.has_path).  Probably nx-cugraph has the most complete [set of algorithms](https://github.com/rapidsai/cugraph/tree/branch-24.06/python/nx-cugraph#supported-algorithms) -- but requires specific gpu hardware.  I believe ArrangoDB has a backend for database graph storage as well.

We also have  [general docs on backends](https://networkx.org/documentation/latest/reference/backends.html).
If you have further questions, would like to contribute to these efforts, or would like to develop your own backend, posting here would be a good place to start and we can direct you to folks and places with more info.

**Comment by dschult** [comment]
I imagine a nx-cython or nx-cpp would be popular if well implemented (whatever that means).   :)

**Comment by Schefflera-Arboricola** [comment]
Hi @GabrielJMS,

Thank you for reaching out with such thoughtful suggestions! It's great to see your interest in optimizing NetworkX for performance-critical tasks, especially given the potential impact on large-scale graph operations.

I think one of NetworkX's core principles is to remain pure Python, which makes it easy to install, maintain, and contribute to. However, this does come at the cost of being slow. Recognizing this, about two years ago, the work on NetworkX dispatching began.

The dispatching system allows users to run a standard NetworkX function with an additional `backend` keyword argument that specifies the backend package containing the alternative implementation. This approach doesn’t alter the code in this(networkx) repository. For example, using `nx.betweenness_centrality(G, backend="parallel")` will run the implementation in the [`nx-parallel`](https://github.com/networkx/nx-parallel) backend package, provided it's installed by the user. NetworkX discovers these backends via the Python [entry_point](https://packaging.python.org/en/latest/specifications/entry-points/). Other than using a `backend` kwarg, you can also dispatch a function to the backend's implementation by setting an environment variable or by passing the backend's graph object(type-based dispatching). You can read more about this in the [NetworkX backend documentation](https://networkx.org/documentation/latest/reference/backends.html). I’d love to hear any feedback you have on these docs or networkx dispatching—we’re always looking to make them more user-friendly and inclusive.

I’ve been heavily involved in developing the [`nx-parallel`](https://github.com/networkx/nx-parallel) backend and contributing to the backend dispatching side of the networkx. Although, I don't have a lot of experience in c++ and cython, I would really like to learn along the way and offer whatever I know. I’d be happy to collaborate and contribute to this. LMK how much involvement do you expect/need from my side :)

Also, feel free to join the weekly [Dispatch meetings](https://scientific-python.org/calendars/networkx.ics) to get more engaged.

Also, most of the dispatching code can be found in the [backends.py](https://github.com/networkx/networkx/blob/main/networkx/utils/backends.py)(the `_dispatchable` decorator which you might have seen on almost all of the algorithms in networkx) and for backend configuration in [configs.py](https://github.com/networkx/networkx/blob/main/networkx/utils/configs.py)

Thank you :)

**Comment by GabrielJMS** [comment]
Really interesting! I wasn't aware of this dispatching mechanism—thanks so much for the information, @dschult  @Schefflera-Arboricola !

In this context, I think creating a Cython backend could be worth exploring. We just need to investigate whether it would outperform the existing backends enough to justify the effort.

Cython is essentially a superset of Python that adds support for optional static type declarations, allowing for efficient compilation into CPython. One of the advantages of Cython is that pure Python code is also valid Cython code. This means we can focus on adding static type declarations only to performance-critical algorithms while leaving the rest of the code as is. However, debugging can be more complex—though I believe PyCharm's professional license offers support for Cython debugging.

When the code is fully Cython (i.e., all variables in a function or inside a for loop are statically typed), Cython also provides integration with OpenMP, making it easier to parallelize loops, for example. More details [here](https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html).

Maybe a Cython backend that leverages parallelism where possible might offer the best performance.
